[Doc] [NextJS] Walkthrough: Using GraphQL in component SSR/SSG

illiakovalenko opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

## Motivation
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update (non-breaking change; modified files are limited to the `/docs` directory)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have read the Contributing guide.
- [ ] My code follows the code style of this project.
- [ ] My code/comments/docs fully adhere to the Code of Conduct.
- [ ] My change is a code change and it requires an update to the documentation.
- [ ] My change is a documentation change and it requires an update to the navigation.

nickwesselman commented on docs/data/routes/docs/nextjs/graphql/component/en.md:
```diff
@@ -0,0 +1,42 @@
+
+Let's go through the steps to execute first GraphQL query in your component. We can use [GraphQL-ConnectedDemo](https://github.com/Sitecore/jss/blob/master/samples/nextjs/src/components/graphql/GraphQL-ConnectedDemo.tsx) example.
+
+1. Define `getStaticProps` or/and `getServerSideProps`. Next steps you should perform inside these functions.
```
To avoid confusion, should clarify that this is in the component itself, as described in the component-level data fetching topic.

illiakovalenko replied:
Mentioned: https://github.com/Sitecore/jss/pull/587/files#diff-f48c7acb9e727a477be3fca37355f49120f17b315fcb9c47421d6c37dff77bd6R31

nickwesselman commented on docs/data/routes/docs/nextjs/graphql/component/en.md:
````diff
@@ -0,0 +1,42 @@
+...
+const graphQLClient = GraphQLClientFactory();
+```
+3. Import needed GraphQL query. [graphql-let](https://github.com/piglovesyou/graphql-let) provides ability to import queries.
````
To me it was confusing what types to use for the query and results, where the names come from, etc. Are there specific `graphql-let` docs we can link to on this?

illiakovalenko replied:
Added note here: https://github.com/Sitecore/jss/pull/587/files#diff-f48c7acb9e727a477be3fca37355f49120f17b315fcb9c47421d6c37dff77bd6R25
And here https://github.com/Sitecore/jss/pull/587/files#diff-f48c7acb9e727a477be3fca37355f49120f17b315fcb9c47421d6c37dff77bd6R15-R16

nickwesselman commented on docs/data/routes/docs/nextjs/graphql/component/en.md:
````diff
@@ -0,0 +1,42 @@
+...
+const graphQLClient = GraphQLClientFactory();
+```
+3. Import needed GraphQL query. [graphql-let](https://github.com/piglovesyou/graphql-let) provides ability to import queries.
````
Should we add details on how the query is expected to be created/defined? And do they need to `graphql:update` after creating it?

Should we add a hint on how to write/debug your query? (with GraphiQL)

anastasiya29 replied:
And what data endpoint this is using

illiakovalenko replied:
Added note: https://github.com/Sitecore/jss/pull/587/files#diff-f48c7acb9e727a477be3fca37355f49120f17b315fcb9c47421d6c37dff77bd6R18

illiakovalenko replied:
Added note about debug, etc: https://github.com/Sitecore/jss/pull/587/files#diff-f48c7acb9e727a477be3fca37355f49120f17b315fcb9c47421d6c37dff77bd6R24-R27

nickwesselman commented on docs/data/routes/docs/nextjs/graphql/component/en.md:
````diff
@@ -0,0 +1,42 @@
+4. Return all needed data for your component. In our case it's `result.data`.
+5. In `render` function access this data using `useComponentProps` hook.
+```ts
+const data = props.rendering.uid
````
Do we need any hints here on rendering the GraphQL data w/ the types generated by `graphql-let`? The need to cast was confusing to me when iterating. Again are there `graphql-let` docs we can link to?

illiakovalenko replied:
Note: https://github.com/Sitecore/jss/pull/587/files#diff-f48c7acb9e727a477be3fca37355f49120f17b315fcb9c47421d6c37dff77bd6R15-R16
Note: https://github.com/Sitecore/jss/pull/587/files#diff-f48c7acb9e727a477be3fca37355f49120f17b315fcb9c47421d6c37dff77bd6R24-R25

anastasiya29 commented on docs/data/routes/docs/nextjs/graphql/component/en.md:
```diff
@@ -0,0 +1,41 @@
+Let's go through the steps to execute first GraphQL query in your component. We can use [GraphQL-ConnectedDemo](https://github.com/Sitecore/jss/blob/master/samples/nextjs/src/components/graphql/GraphQL-ConnectedDemo.tsx) example.
+
+1. Define `getStaticProps` or/and `getServerSideProps`. Next steps you should perform inside these functions.
+2. Create new GraphQL client. `GraphQLClientFactory` is using [apollo-client](https://www.apollographql.com/docs/react) library.
```
Should we provide a sentence about what Apollo is and why we're using this particular library? I see you link to Apollo docs later on in the Walkthrough - I would collect the Apollo background data into a paragraph before the walkthrough

illiakovalenko replied:
Added a little bit about Apollo: https://github.com/Sitecore/jss/pull/587/files#diff-f48c7acb9e727a477be3fca37355f49120f17b315fcb9c47421d6c37dff77bd6R12-R13
We also will have article: `Walkthrough: Using GraphQL on the client`. I expected that Apollo should be described there.

illiakovalenko commented:
@nickwesselman @anastasiya29 Can you check again, please? I tried to cover all questions provided in comments.

illiakovalenko commented:
@nickwesselman @anastasiya29 I removed some part about Introspection, and added link to new article: https://github.com/Sitecore/jss/pull/591. I described this process there.
